src/rtypeconverter.cc

```diff
@@ -300,3 +299,222 @@ void ConvertToProtoBuf::compositeToProtoBuf(SEXP v, std::vector<PlcDataType> &su
         UNPROTECT_PTR(dfcol);
     }
 }
+
+SEXP ConvertToSEXP::arrayToSEXP(const plcontainer::ArrayData &v) {
+    SEXP result, array_dims;
+    int length = v.values_size();
+
+    result = this->allocRVector(v.elementtype(), length);
+    switch (v.elementtype()) {
+        case plcontainer::PlcDataType::LOGICAL: {
+            for (int i = 0; i < length; i++) {
+                if (!v.values()[i].isnull()) {
+                    LOGICAL_DATA(result)[i] = v.values()[i].logicalvalue();
+                } else {
+                    LOGICAL_DATA(result)[i] = NA_LOGICAL;
+                }
+            }
+        } break;
+        case plcontainer::PlcDataType::INT: {
+            for (int i = 0; i < length; i++) {
+                if (!v.values()[i].isnull()) {
+                    INTEGER_DATA(result)[i] = v.values()[i].intvalue();
+                } else {
+                    INTEGER_DATA(result)[i] = NA_INTEGER;
+                }
+            }
+        } break;
+        case plcontainer::PlcDataType::REAL: {
+            for (int i = 0; i < length; i++) {
+                if (!v.values()[i].isnull()) {
+                    NUMERIC_DATA(result)[i] = v.values()[i].realvalue();
+                } else {
+                    NUMERIC_DATA(result)[i] = NA_REAL;
+                }
+            }
+        } break;
+        case plcontainer::PlcDataType::BYTEA: {
+            for (int i = 0; i < length; i++) {
+                if (!v.values()[i].isnull()) {
+                    SET_STRING_ELT(result, i,
+                                   COPY_TO_USER_STRING(v.values()[i].byteavalue().data()));
+                } else {
+                    SET_STRING_ELT(result, i, NA_STRING);
+                }
+            }
+        } break;
+        case plcontainer::PlcDataType::TEXT: {
+            for (int i = 0; i < length; i++) {
+                if (!v.values()[i].isnull()) {
+                    SET_STRING_ELT(result, i,
+                                   COPY_TO_USER_STRING(v.values()[i].stringvalue().c_str()));
+                } else {
+                    SET_STRING_ELT(result, i, NA_STRING);
+                }
+            }
+        } break;
+        default:
+            this->rLog->log(RServerLogLevel::WARNINGS,
```

Comment:
Why not error out here instead of just print WARNING ?

Reply:
ERROR is the unrecoverable issue and will lead rserver to exit. WARNING is the recoverable issue and will not lead rserver to exit.